File: HISTOGRA.py
import sys

def merge(l, r):
    mid = (l+r) / 2
    m = min(histogram[mid], histogram[mid+1])
    if m == histogram[mid]:
        j = mid
    else:
        j = mid+1 
    sol = 0
    i = j
    while(i <= r and m <= histogram[i]):
        sol += m
        i += 1
    i = j-1
    while(i >= 0 and m <= histogram[i]):
        sol += m
        i -= 1
    return sol
        
def solve(l, r):
    if l == r:
        return histogram[l]
    mid = (l+r) / 2
    a = solve(l, mid)
    b = solve(mid + 1, r)
    c = merge(l, r)
    ans = max(a, b, c)
    return ans
    
while 1:
    histogram = map(int, sys.stdin.readline().strip().split())
    N = len(histogram)
    if (N == 1 and histogram[0] == 0):
        break
    print solve(0, N-1)

File: HISTOGRA.cpp
#include<stdio.h>
#include<iostream>
#include<vector>
#include<string.h>
using namespace std;
vector <int> array(10000);

long long int merge(int start, int end){
    int mid = (start+end) >>  1   ;
    int m = array[mid]  ;
    int j = 0   ;
    if (m == array[mid])
        j = mid ;
    else
        j = mid+1   ;
    long long int sol = 0 ;
    int i = j   ;
    while(i <= end && m <= array[i]){
        sol += m    ;
        i ++      ;
    }
    i = j - 1   ;
    while(i >= 0 && m <= array[i]){
        sol += m    ;
        i --      ;
    }
    return sol  ;

}

long long int solve(int start, int end){
    if (start == end)
        return array[start];

    int mid = (start+end) >> 1   ;
    long long int a = solve(start, mid) ;
    long long int b = solve(mid+1, end) ;
    long long int c = merge(start, end)   ;
    return max(max(a, b), c)    ;
}
int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL);
    int size, k;

    while(1){
        scanf("%d", &size)  ;
        if (size == 0)
            return 0;
        for (int i = 0; i < size; i++){
            scanf("%d", &array[i])  ;
            }
        printf("%lld\n", solve(0, size-1))    ;
    }
}
